### DISCUSSION TITLE
Question about gp5 export and instrument channels

### DISCUSSION THREAD

**Discussion by Fristi61** [helge17/tuxguitar#512] [General]
Hello all,

So when trying to export a .tg file with more than 8 instruments to .gp5, I'm running into the familiar issue where it will delete certain instruments and set the associated tracks to use a different instrument instead.

I understand why this happens. There is a limit of 16 channels. Instruments use 2 channels each by default, so after 8 instruments you've normally run out of space.
I also remember from Guitar Pro 5 that this could be worked around by making instruments use 1 channel rather than 2, which saves space (although it means that it won't properly play bends or certain other effects, but that's fine if no such effects are in use).

The strange thing is that I can't seem to get this workaround to stick when exporting to .gp5.

I found that in tuxguitar I can select the instrument, click on a few wrenches and eventually end up on a window that lets me change the MIDI channels from 2 channels to 1 channel.
I would think this is the solution, but, unfortunately, when exporting to .gp5 all my instruments are back to using 2 channels. It seems this information doesn't get carried over?

IIRC Guitar Pro 5 did have the possibility to manually set the effects channel of an instrument to the same number as the main channel. So I would _suppose_ that the .gp5 format does allow for single-channel instruments, but I might be wrong.

So my question is: is this just a currently missing feature of .gp5 export that could easily be added, or is it a more difficult obstacle and do I just have to deal with this limit of the .gp5 format?

Thank you for all your work, I love Tuxguitar.

(My use case for this is when trying to share tabs on Ultimate Guitar. They only accept guitar pro formats, so I am forced to convert my .tg files to .gp5 files for that)

**Comment by guiv42** [comment]
Hi
similar issue mentioned here: #354 

> I found that in tuxguitar I can select the instrument, click on a few wrenches and eventually end up on a window that lets me change the MIDI channels from 2 channels to 1 channel.

Could you please give more precise details about this procedure, so I can investigate?

> IIRC Guitar Pro 5 did have the possibility to manually set the effects channel of an instrument to the same number as the main channel. So I would suppose that the .gp5 format does allow for single-channel instruments, but I might be wrong.

No idea about this. Just looking at [the code](https://github.com/helge17/tuxguitar/blob/bc7bb30cd077b04bb77862c52731ccd1690620b2/common/TuxGuitar-gtp/src/org/herac/tuxguitar/io/gtp/GP5OutputStream.java#L278), it seems 2 channels per track are always written to gp5. I don't know if it's possible to have the same 2 numbers here (that's why I need to understand how you allocate 1 single channel for a track).

And just and additional question: do you need more than 8 *different* instruments? E.g. if you have several tracks using "acoustic guitar" sound, you can associate the same instrument to the different tracks. That would limit the number of channels used (drawback: you would lose the ability to control volume of tracks independently).

**Reply by Fristi61** [reply]
Hello, thanks for your reply.

> similar issue mentioned here: https://github.com/helge17/tuxguitar/issues/354

Hmm, that's funny. While it's similar, I do not actually have that problem in tuxguitar itself. I can add more than 9 instruments and it'll work fine. Only when exporting to .gp5 does the issue occur for me.

> Could you please give more precise details about this procedure, so I can investigate?

Instrument manager (view -> show instruments) -> click on the wrench "settings" to the right of any instrument -> click on the topmost wrench "Edit MIDI receiver" in the new window that opens -> in the newest window it is possible to switch the MIDI channels from 2 to 1 channels

> No idea about this. Just looking at the code it seems 2 channels per track are always written to gp5.

For reference see [this reddit thread](https://www.reddit.com/r/GuitarPro/comments/ssaisp/gp_52_a_new_track_is_mimicking_the_volume_and/) where someone describes how in GP5 the instruments can also be made to use one channel only, which allows for extra instruments.
This is what made me think the .gp5 format might support more instruments if the channels are combined.

> do you need more than 8 different instruments?

I do already set multiple tracks to use the same instrument if they use the same sound program and more or less the same panning.
Lately I've been transcribing a progressive metal album with a rather dense mix of distorted, clean and acoustic guitars (some of which might be panned only to one side), bass, drums, vocals, several types of synth, and sometimes even a koto and/or shamisen for good measure.
Even being frugal with the instruments as you describe, I sometimes end up needing 10 or 11 rather than 8 if I want to get it right.
For now I've worked around the issue by merging the instruments that are not totally important for guitar learning purposes (synths, vocals, some other misc instruments) to just use the same synth instrument but of course it doesn't sound quite right...

**Comment by guiv42** [comment]
> click on the topmost wrench "Edit MIDI receiver" in the new window that opens

That's amazing, there are still features in this app I'm not even aware of. What's your configuration?
In my config I don't see such a control (TuxGuitar 1.6.4, linux opensuse Tumbleweed, settings/sound: TuxGuitar sequencer & gervill)

https://github.com/user-attachments/assets/37846855-50f6-4315-a5d7-b3ecf4f78d59

**Reply by Fristi61** [reply]
TuxGuitar 1.6.4, windows 11, tuxguitar sequencer & tuxguitar synthesizer (default installation settings on windows, I believe)

WOW, actually, switching the MIDI port to gervill solves the problem!
Now I get the same window you do, and here if I set the effect channel to be the same as the normal channel, now it DOES export those settings to .gp5 and I can get more than 8 instruments!

Actually when changing the midi port to gervill I now do get the same issue as in https://github.com/helge17/tuxguitar/issues/354, but, again, it can be worked around by setting the effect channel to the same as the normal channel - and it even exports it to .gp5.



So, the only thing I'm not sure of is why this doesn't work with tuxguitar synthesizer as the MIDI port. It seems with that configuration the "1 channel" option does not get exported to .gp5 the same way - or maybe I misunderstood what the "1 channel" option in the tuxguitar synthesizer does at all. Maybe it's not quite the same as setting the normal channel and effect channel to be the same.

Either way, while I'm not sure what the deal is with the tuxguitar synthesizer, I now have a functioning solution so thank you so much for that :-) Totally didn't expect that changing the midi port could fix this.

**Reply by guiv42** [reply]
> Totally didn't expect that changing the midi port could fix this

I did not expect this either!
> I now have a functioning solution

Good news :)
